### PR TITLE
Ensure efficient and atomic updates of kubeconfigs

### PR DESCRIPTION
**Pull request by johnsca**
The various kubeconfig files are rewritten on every hook even if they don't change. Additionally, due to the use of `kubectl` to write the data using the correct fields, there is a mutli-second window in which clients reading the file could see an empty or partial file. This changes it to only update a given config file if it has actually changed and to do so in an entirely atomic fashion which will not disrupt clients which are actively reading the file.

Fixes [lp:1837930][]

[lp:1837930]: https://bugs.launchpad.net/charm-kubernetes-master/+bug/1837930

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    PYTHONPATH={toxinidir}:{toxinidir}/lib
+    PYTHONBREAKPOINT=ipdb.set_trace
+
+[testenv:lint]
+deps =
+    black
+    flake8
+commands =
+    flake8 {toxinidir}/lib {toxinidir}/tests
+#    black --check {toxinidir}/lib {toxinidir}/tests
```
Comment (kwmonroe) [inline tox.ini]: Did you leave `black` commented out so it didn't muddy the meat of this PR?

**Review by kwmonroe — APPROVED** [review]
LGTM. Lmk if you want to let black reformat before merging.

File: tox.ini (hunk 1 above)
Comment (johnsca) [reply tox.ini]: Indeed I did.  Black reformat applied and `tox.ini` updated.

**Comment by johnsca** [comment]
Backported to stable in 319771f

**Pull request merged**